agent/containermetadata/write_metadata_unix.go

```diff
@@ -0,0 +1,83 @@
+// +build !windows
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package containermetadata
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/engine/dockerclient"
+)
+
+const (
+	mountPoint                     = "/opt/ecs/metadata"
+	ContainerMetadataClientVersion = dockerclient.Version_1_21
+)
+
+// createMetadataFile initializes the metadata file
+func createMetadataFile(metadataDirectoryPath string) error {
+	metadataFilePath := filepath.Join(metadataDirectoryPath, metadataFile)
+	temp, err := ioutil.TempFile(metadataDirectoryPath, "temp_metadata_file")
+	if err != nil {
+		return err
+	}
+	defer temp.Close()
+	err = temp.Chmod(readOnlyPerm)
```

Comment:
I think you're trying to make it rw for the agent, but ro for the task containers.  I don't think this will be effective for the general case here, as the user inside the task container is root by default, same as the agent.  There are circumstances where the user will be different (explicit `USER` directive in the Dockerfile or `User` field in the task definition, or `userns=remap` turned on), but I think that's more rare.